---
title: "VariantStrainSelection"
author: "bwa Alberto Zani"
date: "2024-11-14"
output: html_document
---

# 0. Set Up
```{r setup, include=FALSE}

# Set working directory
knitr::opts_knit$set(root.dir = "/path-to-SIFT-folder/SIFT")

# Load libraries
library(tidyverse)
library(ggplot2)
library(dplyr)
library(patchwork)
library(ggvenn)
library(gridExtra)
library(stringr)
library(stringdist)
library(writexl)
library(readxl)
library(wesanderson)
library(ComplexUpset)
```


# 1. Loading SIFT files
```{r bwa SIFT, include=FALSE}
# Set working directory
setwd("/path-to-SIFT-folder/bwaSIFT")
getwd()
data_dir <- getwd() 

# Get a list of all folders within the main directory
folders <- list.dirs(data_dir, recursive = FALSE)

# Create an empty list to store the data frames
bwaSIFT <- list()

# Create a vector to store sample names and track which ones have already been seen
seen_samples <- c()

# Loop through each folder and load the XLS file inside
for (i in seq_along(folders)) {
  
  # Get the file name of the XLS file inside the folder (without the full path)
  file_name <- list.files(folders[i], pattern = "\\.xls", full.names = TRUE)
  
  # Check if any XLS file was found
  if (length(file_name) > 0) {
    # Use the file name directly as it contains the full path now
    file_path <- file_name[1]  # Take the first file if there are multiple .xls files
    
    # Extract the date and sample number from the file name
    file_basename <- basename(file_path)
    date_sample <- sub("^(\\d{8})Tn5plate\\dCsample(\\d+)_.*$", "\\1sample\\2", file_basename)
    
    # Extract the date and sample number
    date <- substr(date_sample, 1, 8)
    sample_num <- sub(".*sample(\\d+)", "\\1", date_sample)
    
    # Determine suffix based on the date
    suffix <- ifelse(date == "20230619", "1", ifelse(date == "20230901", "2", ""))
    
    # Construct the sample name with suffix
    sample_name <- paste0("sample", sample_num, "_", suffix)
    
    # Add the sample name to the list of seen samples
    seen_samples <- c(seen_samples, sample_name)
    
    # Read the file and store it in the list with a dynamic name based on the folder
    data <- read_tsv(file_path)
    
    # Add "chr" prefix to each value in the CHROM column
    data$CHROM <- paste0("chr", data$CHROM)
    
    # Add the sample column to the dataframe
    data$sample <- sample_name
    
    # Store the data in the list
    bwaSIFT[[paste0(sample_name)]] <- data
  } else {
    # Print a message if no XLS file is found in the current folder
    message("No XLS file found in folder: ", folders[i])
  }
}

rm(data)
```



```{r bowtie SIFT, include=FALSE}
# Set working directory
setwd("/path-to-SIFT-folder/bowtieSIFT")
getwd()
data_dir <- getwd() 

# Get a list of all folders within the main directory
folders <- list.dirs(data_dir, recursive = FALSE)

# Create an empty list to store the data frames
bowtieSIFT <- list()

# Create a vector to store sample names and track which ones have already been seen
seen_samples <- c()

# Loop through each folder and load the XLS file inside
for (i in seq_along(folders)) {
  
  # Get the file name of the XLS file inside the folder (without the full path)
  file_name <- list.files(folders[i], pattern = "\\.xls", full.names = TRUE)
  
  # Check if any XLS file was found
  if (length(file_name) > 0) {
    # Use the file name directly as it contains the full path now
    file_path <- file_name[1]  # Take the first file if there are multiple .xls files
    
    # Extract the date and sample number from the file name
    file_basename <- basename(file_path)
    date_sample <- sub("^(\\d{8})Tn5plate\\dCsample(\\d+)_.*$", "\\1sample\\2", file_basename)
    
    # Extract the date and sample number
    date <- substr(date_sample, 1, 8)
    sample_num <- sub(".*sample(\\d+)", "\\1", date_sample)
    
    # Determine suffix based on the date
    suffix <- ifelse(date == "20230619", "1", ifelse(date == "20230901", "2", ""))
    
    # Construct the sample name with suffix
    sample_name <- paste0("sample", sample_num, "_", suffix)
    
    # Add the sample name to the list of seen samples
    seen_samples <- c(seen_samples, sample_name)
    
    # Read the file and store it in the list with a dynamic name based on the folder
    data <- read_tsv(file_path)
    
    # Add the sample column to the dataframe
    data$sample <- sample_name
    
    # Store the data in the list
    bowtieSIFT[[paste0(sample_name)]] <- data
  } else {
    # Print a message if no XLS file is found in the current folder
    message("No XLS file found in folder: ", folders[i])
  }
}

rm(data)
```



```{r crg SIFT, include=FALSE}
# Set working directory
setwd("/path-to-SIFT-folder/crgSIFT")
getwd()
data_dir <- getwd() 

# Get a list of all folders within the main directory
folders <- list.dirs(data_dir, recursive = FALSE)

crgSIFT <- list()

# Loop through each folder and load the XLS file inside
for (i in seq_along(folders)) {
  # Get the file name of the XLS file inside the folder (without the full path)
  file_name <- list.files(folders [i], pattern = "\\.xls$", full.names = TRUE)
  
  # Check if any XLS file was found
  if (length(file_name) > 0) {
    # Use the file name directly as it contains the full path now
    file_path <- file_name[1]  # Take the first file if there are multiple .xls files
    
    # Get the folder name
    file_basename <- basename(file_path)
    
    # Extract only the part before "_SIFTannotation.xls", removing one character before it
    cleaned_basename <- sub("_SIFTannotations\\.xls$", "", file_basename)
    
    data <- read_tsv(file_path)
    
    # Dynamically assign each dataset to a new variable with sequential names
    crgSIFT[[(paste0(cleaned_basename))]] <- data
  } else {
    # Print a message if no XLS file is found in the current folder
    message("No XLS file found in folder: ", folders[i])
  }
}

rm(data)
```


# 2. Strain aggregation
```{r}
# Load and merge the keys
strain_key <- read_tsv("/path-to-keyfiles/strain_key.tsv")
sample_key <- read_tsv("/path-to-keyfiles/sample_key.tsv")
data_key <- inner_join(strain_key, sample_key, by = "strain_ID")

# Remove the "_" from the "sample_number" column so that it matched the data names from the list
data_key <- data_key %>%
  mutate(sample_number = gsub("_", "", sample_number))

#################
##### bwa #####
#################
for (i in seq_along(bwaSIFT)) {
  # Extract the current sample name without the "_1" and "_2" suffixes
  sample_name_list <- str_remove(names(bwaSIFT)[i], "_[12]$")
  
  # Find the matching haploid_crg info from data_key
  haploid_crg_info <- data_key %>%
    filter(sample_number == sample_name_list) %>%
    pull(haploid_crg)
  
  # Check if haploid_crg_info is found
  if (length(haploid_crg_info) == 1) {
    # Add haploid_crg info as a new column to the dataframe
    bwaSIFT[[i]] <- bwaSIFT[[i]] %>%
      mutate(strain = word(haploid_crg_info, 1, 2)) %>%
      dplyr::select(-sample)
  } else {
    warning(paste("No matching haploid_crg info found for sample:", sample_name_list))
  }
}

###################
##### bowtie #####
###################
for (i in seq_along(bowtieSIFT)) {
  # Extract the current sample name without the "_1" and "_2" suffixes
  sample_name_list <- str_remove(names(bowtieSIFT)[i], "_[12]$")
  
  # Find the matching haploid_crg info from data_key
  haploid_crg_info <- data_key %>%
    filter(sample_number == sample_name_list) %>%
    pull(haploid_crg)
  
  # Check if haploid_crg_info is found
  if (length(haploid_crg_info) == 1) {
    # Add haploid_crg info as a new column to the dataframe
    bowtieSIFT[[i]] <- bowtieSIFT[[i]] %>%
      mutate(strain = word(haploid_crg_info, 1, 2)) %>%
      dplyr::select(-sample)
  } else {
    warning(paste("No matching haploid_crg info found for sample:", sample_name_list))
  }
}

##################
##### crg #####
##################
for (i in seq_along(crgSIFT)) {
  # Get the current sample name
  sample_name_list <- names(crgSIFT)[i]
  
  # Extract the first two words of 'haploid_crg' as 'strain' in data_key_variant
  data_key_variant <- data_key %>%
    mutate(strain = word(haploid_crg, 1, 2))
  
  # Compute similarity scores between the sample name and all strains
  similarity_scores <- stringsim(sample_name_list, data_key_variant$strain)
  
  # Find the strain with the highest similarity score
  best_match_index <- which.max(similarity_scores)
  best_similarity <- similarity_scores[best_match_index]
  
  # Check if the best similarity is above the threshold
  if (best_similarity >= 0.1) {
    best_match_strain <- data_key_variant$strain[best_match_index]
    
    # Add the matched strain to the crgSIFT dataframe
    crgSIFT[[i]] <- crgSIFT[[i]] %>%
      mutate(strain = best_match_strain)
  } else {
    warning(paste("No suitable match found for sample:", sample_name_list))
  }
}

rm(data_key_variant, sample_key, strain_key)
```


# 3. Merge based on the different strains
```{r}
#################
##### bwa #####
#################
# Create the list to store the variants
bwaSIFTVariant <- list()

# Loop over the "bwaSIFT" list
for (i in seq_along(bwaSIFT)) {
  sample <- bwaSIFT[[i]]  # Extract the current sample
  
  strain <- sample$strain[1]  # Get the strain name
  
  if (strain %in% names(bwaSIFTVariant)) {
    # If the strain already exists in the list, append the sample using rbind
    bwaSIFTVariant[[strain]] <- rbind(bwaSIFTVariant[[strain]], sample)
  } else {
    # Otherwise, create a new entry in the list for this strain
    bwaSIFTVariant[[strain]] <- sample
  }
}
  
###################
##### bowtie #####
###################
# Create the list to store the variants
bowtieSIFTVariant <- list()

# Loop over the "bowtieSIFT" list
for (i in seq_along(bowtieSIFT)) {
  sample <- bowtieSIFT[[i]]  # Extract the current sample
  
  strain <- sample$strain[1]  # Get the strain name
  
  if (strain %in% names(bowtieSIFTVariant)) {
    # If the strain already exists in the list, append the sample using rbind
    bowtieSIFTVariant[[strain]] <- rbind(bowtieSIFTVariant[[strain]], sample)
  } else {
    # Otherwise, create a new entry in the list for this strain
    bowtieSIFTVariant[[strain]] <- sample
  }
}

rm(sample)
```


# 4. Venn Plots
```{r venn function for 3 groups, echo=FALSE}
# Corresponding names for each dataset to use in filenames
dataset_names <- c(
  "M22", "BY", "RM", "YPS163", "YJM145", "Clib413", 
  "YJM978", "YJM454", "YPS1009", "I14", "Y10", "PW5", 
  "273614", "YJM981", "CBS2888", "Clib219"
)


# Function to create a Venn diagram with ggvenn for three datasets
fun_venn_diagram <- function(dataset1, dataset2, dataset3, label1 = "Dataset 1", label2 = "Dataset 2", label3 = "Dataset 3", title = "Venn Diagram") {
  
  # Create CHROM_POS columns for each dataset
  dataset1$CHROM_POS <- paste(dataset1$CHROM, dataset1$POS)
  dataset2$CHROM_POS <- paste(dataset2$CHROM, dataset2$POS)
  dataset3$CHROM_POS <- paste(dataset3$CHROM, dataset3$POS)

  # Define unique sets for the Venn diagram
  sets <- list(
    label1 = unique(dataset1$CHROM_POS),
    label2 = unique(dataset2$CHROM_POS),
    label3 = unique(dataset3$CHROM_POS)
  )
  
  # Rename list keys dynamically for labels
  names(sets) <- c(label1, label2, label3)

  # Create a ggvenn plot
  ggvenn_plot <- ggvenn(
    sets,
    fill_color = c("dodgerblue", "goldenrod1", "mediumseagreen"),
    stroke_size = 0.5,
    set_name_size = 5
  ) +
    ggtitle(title)

  return(ggvenn_plot)
}


# Generate individual plots
venn_plots <- list(
  fun_venn_diagram(bwaSIFTVariant[["M22 MatA"]], bowtieSIFTVariant[["M22 MatA"]], crgSIFT[["M22"]], "bwa", "bowtie", "crg", "Venn M22"),
  fun_venn_diagram(bwaSIFTVariant[["M22 MatAlpha"]], bowtieSIFTVariant[["M22 MatAlpha"]], crgSIFT[["M22"]], "bwa", "bowtie", "crg", "Venn M22 Alpha"),
  fun_venn_diagram(bwaSIFTVariant[["BY MatA"]], bowtieSIFTVariant[["BY MatA"]], crgSIFT[["BYa"]], "bwa", "bowtie", "crg", "Venn BY"),
  fun_venn_diagram(bwaSIFTVariant[["RM MatAlpha"]], bowtieSIFTVariant[["RM MatAlpha"]], crgSIFT[["RMx"]], "bwa", "bowtie", "crg", "Venn RM"),
  fun_venn_diagram(bwaSIFTVariant[["YPS163 MatA"]], bowtieSIFTVariant[["YPS163 MatA"]], crgSIFT[["YPS163a"]], "bwa", "bowtie", "crg", "Venn YPS163"),
  fun_venn_diagram(bwaSIFTVariant[["YJM145 MatAlpha"]], bowtieSIFTVariant[["YJM145 MatAlpha"]], crgSIFT[["YJM145x"]], "bwa", "bowtie", "crg", "Venn YJM145"),
  fun_venn_diagram(bwaSIFTVariant[["Clib413 MatA"]], bowtieSIFTVariant[["Clib413 MatA"]], crgSIFT[["CLIB413a"]], "bwa", "bowtie", "crg", "Venn Clib413"),
  fun_venn_diagram(bwaSIFTVariant[["YJM978 MatAlpha"]], bowtieSIFTVariant[["YJM978 MatAlpha"]], crgSIFT[["YJM978x"]], "bwa", "bowtie", "crg", "Venn YJM978"),
  fun_venn_diagram(bwaSIFTVariant[["YJM454 MatA"]], bowtieSIFTVariant[["YJM454 MatA"]], crgSIFT[["YJM454a"]], "bwa", "bowtie", "crg", "Venn YJM454"),
  fun_venn_diagram(bwaSIFTVariant[["YPS1009 MatAlpha"]], bowtieSIFTVariant[["YPS1009 MatAlpha"]], crgSIFT[["YPS1009x"]], "bwa", "bowtie", "crg", "Venn YPS1009"),
  fun_venn_diagram(bwaSIFTVariant[["I14 MatA"]], bowtieSIFTVariant[["I14 MatA"]], crgSIFT[["I14a"]], "bwa", "bowtie", "crg", "Venn I14"),
  fun_venn_diagram(bwaSIFTVariant[["Y10 MatAlpha"]], bowtieSIFTVariant[["Y10 MatAlpha"]], crgSIFT[["Y10x"]], "bwa", "bowtie", "crg", "Venn Y10"),
  fun_venn_diagram(bwaSIFTVariant[["PW5 MatA"]], bowtieSIFTVariant[["PW5 MatA"]], crgSIFT[["PW5a"]], "bwa", "bowtie", "crg", "Venn PW5"),
  fun_venn_diagram(bwaSIFTVariant[["PW5 MatAlpha"]], bowtieSIFTVariant[["PW5 MatAlpha"]], crgSIFT[["PW5a"]], "bwa", "bowtie", "crg", "Venn PW5 Alpha"),
  fun_venn_diagram(bwaSIFTVariant[["273614 MatA"]], bowtieSIFTVariant[["273614 MatA"]], crgSIFT[["273614xa"]], "bwa", "bowtie", "crg", "Venn 273614"),
  fun_venn_diagram(bwaSIFTVariant[["YJM981 MatAlpha"]], bowtieSIFTVariant[["YJM981 MatAlpha"]], crgSIFT[["YJM981x"]], "bwa", "bowtie", "crg", "Venn YJM981"),
  fun_venn_diagram(bwaSIFTVariant[["CBS2888 MatA"]], bowtieSIFTVariant[["CBS2888 MatA"]], crgSIFT[["CBS2888a"]], "bwa", "bowtie", "crg", "Venn CBS2888"),
  fun_venn_diagram(bwaSIFTVariant[["Clib219 MatAlpha"]], bowtieSIFTVariant[["Clib219 MatAlpha"]], crgSIFT[["CLIB219x"]], "bwa", "bowtie", "crg", "Venn Clib219")

)

# Combine plots into a grid using patchwork
final_grid <- wrap_plots(venn_plots, ncol = 6)  # Adjust `ncol` for desired columns

# Display the grid
print(final_grid)

# Plot for thesis
venn_thesis <- list(
  fun_venn_diagram(bwaSIFTVariant[["BY MatA"]], bowtieSIFTVariant[["BY MatA"]], crgSIFT[["BYa"]], "bwa", "bowtie", "crg", "Venn BY"),
  fun_venn_diagram(bwaSIFTVariant[["Y10 MatAlpha"]], bowtieSIFTVariant[["Y10 MatAlpha"]], crgSIFT[["Y10x"]], "bwa", "bowtie", "crg", "Venn Y10"),
  fun_venn_diagram(bwaSIFTVariant[["M22 MatAlpha"]], bowtieSIFTVariant[["M22 MatAlpha"]], crgSIFT[["M22"]], "bwa", "bowtie", "crg", "Venn M22 Alpha"),
  fun_venn_diagram(bwaSIFTVariant[["RM MatAlpha"]], bowtieSIFTVariant[["RM MatAlpha"]], crgSIFT[["RMx"]], "bwa", "bowtie", "crg", "Venn RM")
  )

# Combine plots into a grid using patchwork
grid_thesis <- wrap_plots(venn_thesis, ncol = 2)

# Display the grid
print(grid_thesis)

```


# 4.1 Stats for the Venn
```{r}
# Define strain mapping
strain_info <- list(
  "M22 MatA" = "M22",
  "M22 MatAlpha" = "M22",
  "BY MatA" = "BYa",
  "RM MatAlpha" = "RMx",
  "YPS163 MatA" = "YPS163a",
  "YJM145 MatAlpha" = "YJM145x",
  "Clib413 MatA" = "CLIB413a",
  "YJM978 MatAlpha" = "YJM978x",
  "YJM454 MatA" = "YJM454a",
  "YPS1009 MatAlpha" = "YPS1009x",
  "I14 MatA" = "I14a",
  "Y10 MatAlpha" = "Y10x",
  "PW5 MatA" = "PW5a",
  "PW5 MatAlpha" = "PW5a",
  "273614 MatA" = "273614xa",
  "YJM981 MatAlpha" = "YJM981x",
  "CBS2888 MatA" = "CBS2888a",
  "Clib219 MatAlpha" = "CLIB219x"
)

# Initialize containers
overlap_bwa_crg <- numeric()
overlap_bowtie_crg <- numeric()
overlap_AllThree <- numeric()

for (strain_key in names(strain_info)) {
  crg_key <- strain_info[[strain_key]]
  
  bwa_df <- bwaSIFTVariant[[strain_key]]
  bowtie_df <- bowtieSIFTVariant[[strain_key]]
  crg_df <- crgSIFT[[crg_key]]
  
  if (is.null(bwa_df) || is.null(bowtie_df) || is.null(crg_df)) next
  
  # Create CHROM_POS identifiers
  bwa_pos <- unique(paste(bwa_df$CHROM, bwa_df$POS))
  bowtie_pos <- unique(paste(bowtie_df$CHROM, bowtie_df$POS))
  crg_pos <- unique(paste(crg_df$CHROM, crg_df$POS))
  
  # Compute overlaps using CHROM_POS
  bwa_crg <- length(intersect(bwa_pos, crg_pos)) / length(union(bwa_pos, crg_pos))
  bowtie_crg <- length(intersect(bowtie_pos, crg_pos)) / length(union(bowtie_pos, crg_pos))
  all_three <- length(Reduce(intersect, list(bwa_pos, bowtie_pos, crg_pos))) /
               length(Reduce(union, list(bwa_pos, bowtie_pos, crg_pos)))
  
  # Store results
  overlap_bwa_crg <- c(overlap_bwa_crg, bwa_crg)
  overlap_bowtie_crg <- c(overlap_bowtie_crg, bowtie_crg)
  overlap_AllThree <- c(overlap_AllThree, all_three)
}

# Averages
avg_bwa_crg <- median(overlap_bwa_crg)
avg_bowtie_crg <- median(overlap_bowtie_crg)
avg_all_three <- median(overlap_AllThree)

# Output
cat("Median overlap (Replicate 1 vs crg):", round(avg_bwa_crg, 3), "\n")
cat("Median overlap (Replicate 2 vs crg):", round(avg_bowtie_crg, 3), "\n")
cat("Median overlap (All three):", round(avg_all_three, 3), "\n")

```



# 5. Find the number of deleterious mutations
```{r synonymous and deleterious, echo=FALSE}
#################
##### bwa #####
#################
# Create an empty list to store the filtered samples
bwaSIFTDeleteriousAll <- list()

# Loop through each sample in bwaSIFT
for (sample_name in names(bwaSIFTVariant)) {
  # Access each sample
  sample <- bwaSIFTVariant[[sample_name]]
  
  # Filter the sample
  filtered_sample <- sample %>%
    mutate(DATA = "bwaSIFT")
  
  # Add the filtered sample to the list
  bwaSIFTDeleteriousAll[[sample_name]] <- filtered_sample
}

# Merge all the filtered samples into one dataframe and ensure uniqueness
merged_deleterious_bwa_all <- bind_rows(bwaSIFTDeleteriousAll) %>%
  distinct()  # Remove duplicate rows if there are any

###################
##### bowtie #####
################### 
# Create an empty list to store the filtered samples
bowtieSIFTDeleteriousAll <- list()

# Loop through each sample in "bowtieSIFTVariant"
for (sample_name in names(bowtieSIFTVariant)) {
  # Access each sample
  sample <- bowtieSIFTVariant[[sample_name]]
  
  # Filter the sample 
  filtered_sample <- sample %>%
    mutate(DATA = "bowtieSIFT")
  
  # Add the filtered sample to the list
  bowtieSIFTDeleteriousAll[[sample_name]] <- filtered_sample
}

# Merge all the filtered samples into one dataframe and ensure uniqueness
merged_deleterious_bowtie_all <- bind_rows(bowtieSIFTDeleteriousAll) %>%
  distinct()  # Remove duplicate rows if there are any

##################
##### crg #####
##################
# Create an empty list to store the filtered samples
crgSIFTDeleteriousAll <- list()

# Loop through each sample in bwaSIFT
for (sample_name in names(crgSIFT)) {
  # Access each sample
  sample <- crgSIFT[[sample_name]]
  
  # Filter the sample 
  filtered_sample <- sample %>%
    mutate(DATA = "crgSIFT")
  
  # Add the filtered sample to the list
  crgSIFTDeleteriousAll[[sample_name]] <- filtered_sample
}

merged_deleterious_crg_all <- bind_rows(crgSIFTDeleteriousAll) %>%
    # Remove duplicate rows
  distinct() 

rm(filtered_sample, sample)
```


# 6.Merge the different datasets
```{r}
merged_bwa_bowtie_All <- rbind(merged_deleterious_bwa_all, merged_deleterious_bowtie_all) %>%
  # Filter out unwanted strains
  filter(!strain %in% c("PW5 MatAlpha", "M22 MatAlpha")) %>%
  # Create a unique identifier
  mutate(CHROM_POS = paste(CHROM, POS)) %>%
  # Group by strain and CHROM_POS to check matches within strains
  group_by(strain, CHROM_POS) %>%
  summarise(
    DATA = if_else(n() > 1, "Both", dplyr::first(DATA)),  # Assign "Both" if CHROM_POS matches
    across(everything(), ~ dplyr::first(.)),  # Retain all other columns
    .groups = "drop"
  ) %>%
  # Arrange rows by CHROM and POS
  arrange(CHROM, POS) %>%
  # Perform filtering for proximity or strain differences
  group_by(CHROM) %>%
  ungroup() %>%
  dplyr::select(-CHROM_POS)

# Filter to only include variants found in both datasets
merged_bwa_bowtie_Both_All <- merged_bwa_bowtie_All %>%
  filter(DATA == "Both")

# MERGE bwa, bowtie and crg
merged_bwa_bowtie_crg_All <- rbind(merged_bwa_bowtie_All, merged_deleterious_crg_all) %>%
  # Create a unique identifier
  mutate(CHROM_POS = paste(CHROM, POS)) %>%
  # Group by strain and CHROM_POS to check matches within strains
  group_by(strain, CHROM_POS) %>%
  summarise(
    DATA = if_else(
      n() > 1 & any(DATA %in% c("bowtieSIFT", "bwaSIFT")), 
      "Both",  # If either "bowtieSIFT" or "bwaSIFT" is present, set DATA to "Both"
      if_else(
        n() > 1 & any(DATA == "Both"), 
        "All",  # Assign "All" if the group has >1 row and contains "Both"
        dplyr::first(DATA)  # Otherwise, retain the first DATA value
      )
    ),
    across(everything(), ~ dplyr::first(.)),  # Retain all other columns
    .groups = "drop"
  ) %>%
  # Arrange rows by CHROM and POS
  arrange(CHROM, POS) %>%
  # Perform filtering for proximity or strain differences
  group_by(CHROM) %>%
  ungroup()

# Dataframe with variants from all 3 categories
merged_Common_All <- merged_bwa_bowtie_crg_All %>%
  # Filter relevant DATA categories
  filter(DATA %in% c("All", "bwaSIFT", "bowtieSIFT", "crgSIFT"))

merged_final <- merged_Common_All %>%
  filter(strain %in% c("RM MatAlpha", "BY MatA", "YPS163 MatA", 
                       "Clib219 MatAlpha", "CBS2888 MatA", "PW5 MatA", "Clib413 MatA"))

# Helper function to calculate percentages for plotting
calculate_percentage <- function(df, column) {
  df %>%
    group_by(!!sym(column)) %>%
    summarise(count = n()) %>%
    mutate(percentage = count / sum(count) * 100)
}

# Plots
plots <- list(
  ggplot(calculate_percentage(merged_bwa_bowtie_All, "strain"), 
         aes(x = reorder(strain, count), y = percentage)) +
    geom_bar(stat = "identity", fill = "lightsteelblue") +
    theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
    labs(y = "Percentage", x = "Strain"),
  
  ggplot(calculate_percentage(merged_bwa_bowtie_All, "DATA"), 
         aes(x = reorder(DATA, count), y = percentage)) +
    geom_bar(stat = "identity", fill = "lightgoldenrod3") +
    theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
    labs(y = "Percentage", x = "DATA"),
  
  ggplot(calculate_percentage(merged_bwa_bowtie_crg_All, "strain"), 
         aes(x = reorder(strain, count), y = percentage)) +
    geom_bar(stat = "identity", fill = "lightsteelblue") +
    theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
    labs(y = "Percentage", x = "Strain"),
  
  ggplot(calculate_percentage(merged_bwa_bowtie_crg_All, "DATA"), 
         aes(x = reorder(DATA, count), y = percentage)) +
    geom_bar(stat = "identity", fill = "lightgoldenrod3") +
    theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
    labs(y = "Percentage", x = "DATA")
)

# Combine plots into a grid
grid <- wrap_plots(plots, ncol = 2)
print(grid)
```


# 7. Merge based on key genes for metabolic pathways
```{r}
key_genes <- read.delim("/path-to-keyfiles/514_MitoGenes_then_55Glycogluco_Genes.txt")

colnames(key_genes) <- key_genes[2,]
key_genes <- key_genes[-c(1,2),]
key_genes <- key_genes %>%
  rename(GENE_ID = ORF) %>%
  dplyr::select(GENE_ID, Symbol, Chr, StartBp, EndBP) %>%
  filter(GENE_ID != "" & !is.na(GENE_ID))

# Convert StartBp and EndBP to numeric in key_genes
key_genes <- key_genes %>%
  mutate(StartBp = as.numeric(StartBp), EndBP = as.numeric(EndBP))

# Preprocess `CHROM` column in `matching_merged`
matching_merged <- merged_Common_All %>%
  mutate(CHROM = gsub("^chr", "", CHROM, ignore.case = TRUE))

filtered_matching_merged <- matching_merged %>%
  filter(strain %in% c("RM MatAlpha", "BY MatA", "YPS163 MatA", 
                       "Clib219 MatAlpha", "CBS2888 MatA", "PW5 MatA", "Clib413 MatA")) %>%
  inner_join(key_genes, by = c("CHROM" = "Chr"), relationship = "many-to-many") %>%  # Match rows based on chromosome
  dplyr::select(-SIFT_PREDICTION) %>%
  filter(POS >= (StartBp - 1000) & POS <= (EndBP + 1000)) %>% # Filter for positions within range
  mutate(Position_Category = case_when(
    POS >= StartBp & POS <= EndBP ~ "Within Gene",
    (POS >= (StartBp - 500) & POS < StartBp) | (POS > EndBP & POS <= (EndBP + 500)) ~ "0–500 bp Flanking",
    (POS >= (StartBp - 1000) & POS < (StartBp - 500)) | (POS > (EndBP + 500) & POS <= (EndBP + 1000)) ~ "501–1000 bp Flanking",
    TRUE ~ "Outside Range" # Fallback for any unexpected values
  ))

# Group by position category and by variant type
frameshift_count <- filtered_matching_merged %>%
  group_by(Position_Category, VARIANT_TYPE) %>%
  summarize(count = n(), .groups = "drop") %>%
  group_by(VARIANT_TYPE) %>%
  mutate(proportion = count / sum(count))  # Calculate proportion within each VARIANT_TYPE

ggplot(frameshift_count, aes(x = VARIANT_TYPE, y = proportion, fill = Position_Category)) +
  geom_bar(stat = "identity") + # Use stat = "identity" to plot proportions directly
  coord_flip() +
  ylab("Proportion")  # Label the y-axis as 'Proportion'

# Count removed rows
removed_rows_count <- nrow(matching_merged) - nrow(filtered_matching_merged)
print(removed_rows_count)

length(unique(filtered_matching_merged$CHROM_POS))
length(unique(filtered_matching_merged$GENE_ID.y))

sum(duplicated(key_genes$GENE_ID))

```


# 8. Visualization
```{r}

# Step 1: Create SNP column
filtered_matching_merged$SNP <- paste(
  filtered_matching_merged$CHROM,
  filtered_matching_merged$POS,
  filtered_matching_merged$REF_ALLELE,
  filtered_matching_merged$ALT_ALLELE
)

# Step 2: Select SNP and strain
FinalSNPsData <- filtered_matching_merged %>%
  select(SNP, strain, Position_Category)

# Create the wide format needed for upset plot
t_wide <- FinalSNPsData %>%
  group_by(SNP, strain, Position_Category) %>%
  summarize(n = n(), .groups = "drop") %>%  # Manually summarize to count occurrences of SNP
  mutate(
    win = 1,  # Boolean for presence
    Total = sum(n)  # Total occurrences per SNP
  ) %>%
  pivot_wider(
    id_cols = c(SNP, Position_Category, Total),  # Keep SNP, Position_Category, and Total columns
    names_from = strain,  # Wide format for strains
    values_from = win,  # Fill from win column
    values_fill = list(win = 0)  # Fill missing cells with 0
  ) %>%
  data.frame()


colnames(t_wide) <- gsub("\\..*", "", colnames(t_wide))

# Define the sets (7 strains)
sets <- c(
  "RM", 
  "BY", 
  "YPS163", 
  "Clib219", 
  "CBS2888", 
  "PW5", 
  "Clib413"
)

# View the result
head(t_wide)

# Define the palette
pallette <- wes_palette("Chevalier1")

# Manually map the colors
color_map <- c(
  'Within Gene' = pallette[1],
  '0–500 bp Flanking' = pallette[2], 
  '501–1000 bp Flanking' = pallette[3]
)

# Create the upset plot with correct color mapping
upset_plot <- ComplexUpset::upset(
  t_wide,
  intersect = sets,
  name = "Strains",
  sort_intersections_by = c("degree", "cardinality"),
  min_size = 5,
  base_annotations = list(
    'Intersection size' = intersection_size(
      counts = FALSE,
      mapping = aes(fill = Position_Category)
    ) +
      scale_fill_manual(values = color_map, name = "Relative position to the gene")  # Apply the manual color mapping here
  ),
  set_sizes = (
    upset_set_size(
      geom = geom_bar(
        aes(fill = Position_Category),  # Map fill to Position_Category
        width = 0.8
      ),
      position = 'right'
    ) +
      scale_fill_manual(values = color_map)  # Apply the same color mapping
  ),
  width_ratio = 0.1
)


# Print the number of SNPs per position in the gene
table(FinalSNPsData$Position_Category)

# Check how many unique per position category
print(upset_plot)
unique_SNP <- FinalSNPsData %>%
  distinct(SNP, .keep_all = TRUE)

table(unique_SNP$Position_Category)

```


# 8.1 Stats for the thesis regarding the plot
```{r}

# Total SNP count per strain
strain_snp_counts <- FinalSNPsData %>%
  group_by(strain) %>%
  summarise(total_snps = n())

# Count how many strains each SNP appears in
snp_strain_counts <- FinalSNPsData %>%
  distinct(SNP, strain) %>%
  group_by(SNP) %>%
  summarise(num_strains = n())

# Join back to get num_strains for each SNP-strain pair
snp_with_counts <- FinalSNPsData %>%
  left_join(snp_strain_counts, by = "SNP")

# Unique SNPs = SNPs that appear in only one strain
unique_snp_percentages <- snp_with_counts %>%
  filter(num_strains == 1) %>%
  group_by(strain) %>%
  summarise(unique_snps = n()) %>%
  left_join(strain_snp_counts, by = "strain") %>%
  mutate(unique_percent = 100 * unique_snps / total_snps)

# Shared SNPs = SNPs that appear in multiple strains
shared_snp_percentages <- snp_with_counts %>%
  filter(num_strains > 1) %>%
  group_by(strain) %>%
  summarise(shared_snps = n()) %>%
  left_join(strain_snp_counts, by = "strain") %>%
  mutate(shared_percent = 100 * shared_snps / total_snps)

# Percentages of position categories
position_percentages <- unique_SNP %>%
     count(Position_Category) %>%
     mutate(percentage = 100 * n / sum(n))

position_percentages
```

